File: docs/examples/python/basic_client.py
#!/usr/bin/env python3
"""
Basic Python client for Synology ActiveBackup for Business API

Usage:
    python basic_client.py
"""

import requests
import json
from typing import Dict, Any, Optional


class ActiveBackupClient:
    """Client for Synology ActiveBackup for Business API"""
    
    def __init__(self, host: str, username: str, password: str, use_https: bool = False):
        """
        Initialize the client
        
        Args:
            host: NAS hostname or IP address
            username: Admin username
            password: Admin password
            use_https: Use HTTPS instead of HTTP (default: False)
        """
        self.host = host
        self.username = username
        self.password = password
        protocol = "https" if use_https else "http"
        port = 5001 if use_https else 5000
        self.base_url = f"{protocol}://{host}:{port}/webapi"
        self.sid: Optional[str] = None
        
    def login(self) -> bool:
        """
        Login and obtain session ID
        
        Returns:
            True if login successful, False otherwise
        """
        response = requests.post(
            f"{self.base_url}/auth.cgi",
            data={
                "api": "SYNO.API.Auth",
                "version": "6",
                "method": "login",
                "account": self.username,
                "passwd": self.password,
                "session": "ActiveBackup",
                "format": "sid"
            },
            verify=False  # Disable SSL verification for self-signed certs
        )
        
        data = response.json()
        if data.get("success"):
            self.sid = data["data"]["sid"]
            print(f"✅ Logged in successfully")
            return True
        else:
            error = data.get("error", {})
            print(f"❌ Login failed: Error {error.get('code')}")
            return False
    
    def logout(self) -> bool:
        """
        Logout and invalidate session
        
        Returns:
            True if logout successful, False otherwise
        """
        if not self.sid:
            return True
            
        response = requests.post(
            f"{self.base_url}/auth.cgi",
            data={
                "api": "SYNO.API.Auth",
                "version": "6",
                "method": "logout",
                "session": "ActiveBackup",
                "_sid": self.sid
            },
            verify=False
        )
        
        data = response.json()
        if data.get("success"):
            print(f"✅ Logged out successfully")
            self.sid = None
            return True
        return False
    
    def call_api(self, api: str, method: str, version: str = "1", **params) -> Dict[str, Any]:
        """
        Call an API method
        
        Args:
            api: API name (e.g., "SYNO.ActiveBackup.Device")
            method: Method name (e.g., "list")
            version: API version (default: "1")
            **params: Additional parameters
            
        Returns:
            API response as dictionary
        """
        if not self.sid:
            raise Exception("Not logged in. Call login() first.")
        
        params.update({
            "api": api,
            "version": version,
            "method": method,
            "_sid": self.sid
        })
        
        response = requests.get(
            f"{self.base_url}/entry.cgi",
            params=params,
            verify=False
        )
        
        return response.json()
    
    def get_overview(self) -> Dict[str, Any]:
        """Get backup overview"""
        return self.call_api("SYNO.ActiveBackup.Overview", "get")
    
    def list_devices(self) -> Dict[str, Any]:
        """List all backup devices"""
        return self.call_api("SYNO.ActiveBackup.Device", "list")
    
    def list_tasks(self) -> Dict[str, Any]:
        """List all backup tasks"""
        return self.call_api("SYNO.ActiveBackup.Task", "list")
    
    def get_storage_info(self) -> Dict[str, Any]:
        """Get storage information"""
        return self.call_api("SYNO.ActiveBackup.Storage", "get")


def main():
    """Example usage"""
    
    # Configuration
    NAS_HOST = "192.168.1.100"
    USERNAME = "admin"
    PASSWORD = "your_password"
    
    # Create client
    client = ActiveBackupClient(NAS_HOST, USERNAME, PASSWORD)
    
    try:
        # Login
        if not client.login():
            return
        
        # Get overview
        print("\n📊 Backup Overview:")
        overview = client.get_overview()
        if overview.get("success"):
            print(json.dumps(overview["data"], indent=2))
        
        # List devices
        print("\n💻 Backup Devices:")
        devices = client.list_devices()
        if devices.get("success"):
            device_list = devices["data"].get("devices", [])
            print(f"Total devices: {len(device_list)}")
            for device in device_list[:5]:  # Show first 5
                print(f"  - {device.get('name', 'Unknown')}")
        
        # List tasks
        print("\n📋 Backup Tasks:")
        tasks = client.list_tasks()
        if tasks.get("success"):
            task_list = tasks["data"].get("tasks", [])
            print(f"Total tasks: {len(task_list)}")
            for task in task_list[:5]:  # Show first 5
                print(f"  - {task.get('name', 'Unknown')}")
        
    finally:
        # Always logout
        client.logout()


if __name__ == "__main__":
    main()

File: docs/examples/javascript/basic-client.js
#!/usr/bin/env node
/**
 * Basic Node.js client for Synology ActiveBackup for Business API
 *
 * Usage:
 *   node basic-client.js
 *
 * Requirements:
 *   npm install axios
 */

const axios = require('axios');
const https = require('https');

class ActiveBackupClient {
    /**
     * Initialize the client
     *
     * @param {string} host - NAS hostname or IP address
     * @param {string} username - Admin username
     * @param {string} password - Admin password
     * @param {boolean} useHttps - Use HTTPS instead of HTTP (default: false)
     */
    constructor(host, username, password, useHttps = false) {
        this.host = host;
        this.username = username;
        this.password = password;
        const protocol = useHttps ? 'https' : 'http';
        const port = useHttps ? 5001 : 5000;
        this.baseUrl = `${protocol}://${host}:${port}/webapi`;
        this.sid = null;

        // Create axios instance with SSL verification disabled for self-signed certs
        this.axios = axios.create({
            httpsAgent: new https.Agent({
                rejectUnauthorized: false
            })
        });
    }

    /**
     * Login and obtain session ID
     *
     * @returns {Promise<boolean>} True if login successful
     */
    async login() {
        try {
            const response = await this.axios.post(`${this.baseUrl}/auth.cgi`, null, {
                params: {
                    api: 'SYNO.API.Auth',
                    version: '6',
                    method: 'login',
                    account: this.username,
                    passwd: this.password,
                    session: 'ActiveBackup',
                    format: 'sid'
                }
            });

            const data = response.data;
            if (data.success) {
                this.sid = data.data.sid;
                console.log('✅ Logged in successfully');
                return true;
            } else {
                const error = data.error || {};
                console.error(`❌ Login failed: Error ${error.code}`);
                return false;
            }
        } catch (error) {
            console.error('❌ Login error:', error.message);
            return false;
        }
    }

    /**
     * Logout and invalidate session
     *
     * @returns {Promise<boolean>} True if logout successful
     */
    async logout() {
        if (!this.sid) {
            return true;
        }

        try {
            const response = await this.axios.post(`${this.baseUrl}/auth.cgi`, null, {
                params: {
                    api: 'SYNO.API.Auth',
                    version: '6',
                    method: 'logout',
                    session: 'ActiveBackup',
                    _sid: this.sid
                }
            });

            if (response.data.success) {
                console.log('✅ Logged out successfully');
                this.sid = null;
                return true;
            }
            return false;
        } catch (error) {
            console.error('❌ Logout error:', error.message);
            return false;
        }
    }

    /**
     * Call an API method
     *
     * @param {string} api - API name (e.g., "SYNO.ActiveBackup.Device")
     * @param {string} method - Method name (e.g., "list")
     * @param {string} version - API version (default: "1")
     * @param {Object} params - Additional parameters
     * @returns {Promise<Object>} API response
     */
    async callApi(api, method, version = '1', params = {}) {
        if (!this.sid) {
            throw new Error('Not logged in. Call login() first.');
        }

        try {
            const response = await this.axios.get(`${this.baseUrl}/entry.cgi`, {
                params: {
                    api,
                    version,
                    method,
                    _sid: this.sid,
                    ...params
                }
            });

            return response.data;
        } catch (error) {
            console.error('❌ API call error:', error.message);
            throw error;
        }
    }

    /**
     * Get backup overview
     */
    async getOverview() {
        return this.callApi('SYNO.ActiveBackup.Overview', 'get');
    }

    /**
     * List all backup devices
     */
    async listDevices() {
        return this.callApi('SYNO.ActiveBackup.Device', 'list');
    }

    /**
     * List all backup tasks
     */
    async listTasks() {
        return this.callApi('SYNO.ActiveBackup.Task', 'list');
    }

    /**
     * Get storage information
     */
    async getStorageInfo() {
        return this.callApi('SYNO.ActiveBackup.Storage', 'get');
    }
}

/**
 * Example usage
 */
async function main() {
    // Configuration
    const NAS_HOST = '192.168.1.100';
    const USERNAME = 'admin';
    const PASSWORD = 'your_password';

    // Create client
    const client = new ActiveBackupClient(NAS_HOST, USERNAME, PASSWORD);

    try {
        // Login
        if (!await client.login()) {
            return;
        }

        // Get overview
        console.log('\n📊 Backup Overview:');
        const overview = await client.getOverview();
        if (overview.success) {
            console.log(JSON.stringify(overview.data, null, 2));
        }

        // List devices
        console.log('\n💻 Backup Devices:');
        const devices = await client.listDevices();
        if (devices.success) {
            const deviceList = devices.data.devices || [];
            console.log(`Total devices: ${deviceList.length}`);
            deviceList.slice(0, 5).forEach(device => {
                console.log(`  - ${device.name || 'Unknown'}`);
            });
        }

        // List tasks
        console.log('\n📋 Backup Tasks:');
        const tasks = await client.listTasks();
        if (tasks.success) {
            const taskList = tasks.data.tasks || [];
            console.log(`Total tasks: ${taskList.length}`);
            taskList.slice(0, 5).forEach(task => {
                console.log(`  - ${task.name || 'Unknown'}`);
            });
        }

        // Get storage info
        console.log('\n💾 Storage Information:');
        const storage = await client.getStorageInfo();
        if (storage.success) {
            console.log(JSON.stringify(storage.data, null, 2));
        }

    } catch (error) {
        console.error('Error:', error.message);
    } finally {
        // Always logout
        await client.logout();
    }
}

// Run if executed directly
if (require.main === module) {
    main().catch(console.error);
}

// Export for use as module
module.exports = ActiveBackupClient;

